Fixed PS-7275 (Status variable Innodb_checkpoint_max_age missed from …

altmannmarcelo opened the pull request:
…version 8.0)

https://jira.percona.com/browse/PS-7275

Innodb_checkpoint_max_age is the same value as log_capacity and
has been removed from 8.0. There is no way to currently get
log_capacity value from a user point of view.
Adding back Innodb_checkpoint_max_age as a status variable and
as part of SHOW ENGINE INNODB STATUS.

altmannmarcelo commented:
https://ps80.cd.percona.com/job/percona-server-8.0-param/773/

altmannmarcelo commented:
Note to reviewers: We use `log_get_free_check_capacity(log)` at https://github.com/percona/percona-server/blob/Percona-Server-8.0.20-11/storage/innobase/log/log0chkp.cc#L1236 to validate if we need to wait for redo log space

percona-ysorokin commented on storage/innobase/log/log0chkp.cc:
```diff
@@ -1189,6 +1189,8 @@ void log_calc_max_ages(log_t &log) {
 
   const lsn_t limit = log.lsn_capacity_for_free_check;
 
+  log.max_checkpoint_age.store(log_get_free_check_capacity(log));
```
[optional] There is actually no need to use store()/load() on std::atomic if you are using default parameter for `memory_order` (sequentially consistent). operator = / operator T overloads work exactly the same way
```suggestion
  log.max_checkpoint_age = log_get_free_check_capacity(log);
```

percona-ysorokin commented on storage/innobase/log/log0chkp.cc:
```diff
@@ -1199,6 +1201,9 @@ void log_calc_max_ages(log_t &log) {
   /* Round limits to equal size of OS_FILE_LOG_BLOCK_SIZE, not to risk
   inproper lsn computations (we.g. value % OS_FILE_LOG_BLOCK_SIZE == 1) */
 
+  log.max_checkpoint_age.store(
```
Why do we update the same var `log.max_checkpoint_age` within the same function twice (with unaligned and with aligned value)?

satya-bodapati replied:
Oh, you asked this question already ;) nice

percona-ysorokin commented on storage/innobase/log/log0chkp.cc:
```diff
@@ -1232,6 +1237,8 @@ static void log_update_limits_low(log_t &log) {
   ut_ad(log_limits_mutex_own(log));
 
   const lsn_t log_capacity = log_get_free_check_capacity(log);
+  if (log_capacity != log.max_checkpoint_age.load())
+    log.max_checkpoint_age.store(log_capacity);
```
Shouldn't it be just
```suggestion
   log.max_checkpoint_age. = log_capacity;
```

percona-ysorokin commented on storage/innobase/log/log0log.cc:
```diff
@@ -901,12 +901,14 @@ void log_print(const log_t &log, FILE *file) {
           "\n"
           "Checkpoint age        " LSN_PF "\n",
-          log_sys->max_checkpoint_age_async,
+          log_sys->max_checkpoint_age.load(), log_sys->max_checkpoint_age_async,
```
[optional]
```suggestion
          log_sys->max_checkpoint_age, log_sys->max_checkpoint_age_async,
```

percona-ysorokin commented on storage/innobase/log/log0log.cc:
```diff
@@ -928,8 +930,10 @@ void log_print(const log_t &log, FILE *file) {
     checkpoint age */
     fprintf(file,
             "Log tracking enabled\n"
-            "Log tracked up to   " LSN_PF "\n",
-            log_sys->tracked_lsn.load());
+            "Log tracked up to   " LSN_PF
+            "\n"
+            "Max tracked LSN age " LSN_PF "\n",
+            log_sys->tracked_lsn.load(), log_sys->max_checkpoint_age.load());
```
[optional]
```suggestion
            log_sys->tracked_lsn.load(), log_sys->max_checkpoint_age);
```

satya-bodapati replied:
Can you please print sample output as  github comment?

percona-ysorokin commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -1746,6 +1746,8 @@ void srv_export_innodb_status(void) {
   export_vars.innodb_checkpoint_age =
       (log_get_lsn(*log_sys) - log_sys->last_checkpoint_lsn);
+
+  export_vars.innodb_checkpoint_max_age = log_sys->max_checkpoint_age.load();
```
[optional]
```suggestion
  export_vars.innodb_checkpoint_max_age = log_sys->max_checkpoint_age;
```

satya-bodapati commented on storage/innobase/log/log0chkp.cc:
```diff
@@ -1199,6 +1201,9 @@ void log_calc_max_ages(log_t &log) {
   /* Round limits to equal size of OS_FILE_LOG_BLOCK_SIZE, not to risk
   inproper lsn computations (we.g. value % OS_FILE_LOG_BLOCK_SIZE == 1) */
 
+  log.max_checkpoint_age.store(
+      ut_uint64_align_up(log.max_checkpoint_age, OS_FILE_LOG_BLOCK_SIZE));
+
```
Why do we need to set this variable two times?

satya-bodapati commented on storage/innobase/include/log0types.h:
```diff
@@ -692,6 +692,10 @@ struct alignas(ut::INNODB_CACHE_LINE_SIZE) log_t {
   if lag between oldest_lsn and checkpoint_lsn exceeds max_checkpoint_lag. */
   lsn_t max_checkpoint_age_async;
 
+  /** When checkpoint age exceeds this value, user thread needs
+  to wait. The check is performed when a new query step is started. */
+  atomic_lsn_t max_checkpoint_age;
```
Do u know where this variable is used? log_free_capacity(log) used may be? better to use this max_checkpoint_age variable then? 

And afaiu, this variable value, the max age doesn't change runtime right? (and it fixed value based on redo log file size?)

satya-bodapati commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -1746,6 +1746,8 @@ void srv_export_innodb_status(void) {
   }
   export_vars.innodb_checkpoint_age =
       (log_get_lsn(*log_sys) - log_sys->last_checkpoint_lsn);
+
+  export_vars.innodb_checkpoint_max_age = log_sys->max_checkpoint_age.load();
   ibuf_export_ibuf_status(&export_vars.innodb_ibuf_free_list,
```
Can you please print sample output as  github comment?

altmannmarcelo commented:
@percona-ysorokin / @satya-bodapati - Satya asked to add some extra information on SEIS LOG output (Log size, number of logs, total log size). I've added it as part of my new commit. Also, prior to this commit, output of SEIS was not aligned on log part:

```
---
LOG
---
Log sequence number          608039671
Log buffer assigned up to    608039671
Log buffer completed up to   608039029
Log written up to            608039029
Log flushed up to            608039029
Added dirty pages up to      608039551
Pages flushed up to          591179191
Last checkpoint at           591179191
Max checkpoint age    80576000
Checkpoint age target 83374592
Modified age no less than 16860480
Checkpoint age        16860480
```

I have fixed it too and new LOG section of SEIS look like below:

```
---
LOG
---
Log sequence number          744435847
Log buffer assigned up to    744435847
Log buffer completed up to   744435621
Log written up to            744435621
Log flushed up to            744435621
Added dirty pages up to      744435805
Pages flushed up to          743321252
Last checkpoint at           743321252
Max checkpoint age           80576000
Checkpoint age target        83374592
Modified age no less than    1114595
Checkpoint age               1114595
Log size                     50331648
Number of logs               2
Log total capacity           100663296
```

Output of SHOW GLOBAL STATUS:

```
mysql> SHOW GLOBAL STATUS LIKE '%checkpoint%';
+----------------------------+-----------+
| Variable_name              | Value     |
+----------------------------+-----------+
| Innodb_checkpoint_age      | 17055090  |
| Innodb_checkpoint_max_age  | 80576000  |
| Innodb_lsn_last_checkpoint | 123223358 |
+----------------------------+-----------+
```

satya-bodapati commented:
> @percona-ysorokin / @satya-bodapati - Satya asked to add some extra information on SEIS LOG output (Log size, number of logs, total log size). I've added it as part of my new commit. Also, prior to this commit, output of SEIS was not aligned on log part:
> 
> ```
> ---
> LOG
> ---
> Log sequence number          608039671
> Log buffer assigned up to    608039671
> Log buffer completed up to   608039029
> Log written up to            608039029
> Log flushed up to            608039029
> Added dirty pages up to      608039551
> Pages flushed up to          591179191
> Last checkpoint at           591179191
> Max checkpoint age    80576000
> Checkpoint age target 83374592
> Modified age no less than 16860480
> Checkpoint age        16860480
> ```
> 
> I have fixed it too and new LOG section of SEIS look like below:
> 
> ```
> ---
> LOG
> ---
> Log sequence number          744435847
> Log buffer assigned up to    744435847
> Log buffer completed up to   744435621
> Log written up to            744435621
> Log flushed up to            744435621
> Added dirty pages up to      744435805
> Pages flushed up to          743321252
> Last checkpoint at           743321252
> Max checkpoint age           80576000
> Checkpoint age target        83374592
> Modified age no less than    1114595
> Checkpoint age               1114595
> Log size                     50331648
> Number of logs               2
> Log total capacity           100663296
> ```
> 
> Output of SHOW GLOBAL STATUS:
> 
> ```
> mysql> SHOW GLOBAL STATUS LIKE '%checkpoint%';
> +----------------------------+-----------+
> | Variable_name              | Value     |
> +----------------------------+-----------+
> | Innodb_checkpoint_age      | 17055090  |
> | Innodb_checkpoint_max_age  | 80576000  |
> | Innodb_lsn_last_checkpoint | 123223358 |
> +----------------------------+-----------+
> ```

@altmannmarcelo looks good. Minor suggestion (rerrangement and name change):
```
---
LOG
---
Log sequence number          744435847
Log buffer assigned up to    744435847
Log buffer completed up to   744435621
Log written up to            744435621
Log flushed up to            744435621
Added dirty pages up to      744435805
Pages flushed up to          743321252
Last checkpoint at           743321252
Checkpoint age target        83374592
Modified age no less than    1114595
Checkpoint age               1114595
Max checkpoint age           80576000
Number of logs               2
Log size                     50331648
Log total size               100663296
```

satya-bodapati reviewed (changes requested):
See suggestion submitted as comment

altmannmarcelo commented:
@satya-bodapati thanks for your suggestion. It looks more logical now. See full output after the changes:

```
---
LOG
---
Log sequence number          17902927
Log buffer assigned up to    17902927
Log buffer completed up to   17902927
Log written up to            17902927
Log flushed up to            17902927
Added dirty pages up to      17902927
Pages flushed up to          17902877
Last checkpoint at           17902877
Checkpoint age target        83374592
Modified age no less than    50
Checkpoint age               50
Max checkpoint age           80576000
Number of logs               2
Log size                     50331648
Log total capacity           100663296
28 log i/o's done, 0.12 log i/o's/second
Log tracking enabled
Log tracked up to            17902877
Max tracked LSN age          80576000
```

satya-bodapati commented on storage/innobase/log/log0log.cc:
```diff
@@ -901,14 +903,24 @@ void log_print(const log_t &log, FILE *file) {
+          "\n"
+          "Log size                     " LSN_PF
+          "\n"
+          "Log total capacity           " LSN_PF "\n",
```
Log total size

altmannmarcelo replied:
Done

satya-bodapati commented on storage/innobase/log/log0log.cc:
```diff
@@ -928,8 +940,10 @@ void log_print(const log_t &log, FILE *file) {
-            log_sys->tracked_lsn.load());
+            "Log tracked up to            " LSN_PF
+            "\n"
+            "Max tracked LSN age          " LSN_PF "\n",
+            log_sys->tracked_lsn.load(), max_checkpoint_age);
```
Do we need this duplicated print? Did it exist in 5.7?

altmannmarcelo replied:
@satya-bodapati yes, added here for consistency - https://github.com/percona/percona-server/blob/Percona-Server-5.7.31-34/storage/innobase/log/log0log.cc#L2918

satya-bodapati replied:
ok, lets keep it then

satya-bodapati reviewed (approved):
LGTM!

percona-ysorokin reviewed (approved):
LGTM
